src/services/safe_border.py

```diff
@@ -0,0 +1,134 @@
+        bunker_start_epoch = get_epoch_by_timestamp(bunker_start_timestamp) # 29750
+
+        bunker_start_border_epoch = bunker_start_epoch - NEW_REQUESTS_BORDER # 29742
+        earliest_allowable_epoch = get_epoch_by_slot(blockstamp.slot_number) - MAX_NEGATIVE_REBASE_BORDER # 29714
+
+        return max(earliest_allowable_epoch, bunker_start_border_epoch)
+
+    def get_associated_slashings_border_epoch(self, blockstamp: BlockStamp):
+        earliest_slashed_epoch = self.get_earliest_slashed_epoch_among_incomplete_slashings(blockstamp)
+
+        if earliest_slashed_epoch is None:
+            return get_epoch_by_slot(blockstamp.slot_number) - NEW_REQUESTS_BORDER
+        
+        # should NEW_REQUESTS_BORDER be here?
+        return earliest_slashed_epoch - NEW_REQUESTS_BORDER
+
+    def get_earliest_slashed_epoch_among_incomplete_slashings(self, blockstamp: BlockStamp):
+        validators = self.get_lido_validators(blockstamp)       
+        validators_slashed = filter_slashed_validators(validators)
+        validators_slashed_non_withdrawn = filter_non_withdrawable_validators(validators_slashed, blockstamp.slot_number)
+
+        if len(validators_slashed_non_withdrawn) == 0:
+            return None
+
+        validators_with_earliest_exit_epoch = self.get_validators_with_earliest_exit_epoch(validators_slashed_non_withdrawn)
+        first_validator_with_earliest_exit_epoch = validators_with_earliest_exit_epoch[0]
+        earliest_slashed_epoch = self.calc_validator_slashed_epoch_from_state(first_validator_with_earliest_exit_epoch)
+        if earliest_slashed_epoch is not None:
+            return earliest_slashed_epoch
+        
+        earliest_slashed_epoch = self.find_earliest_slashed_epoch(validators_with_earliest_exit_epoch, blockstamp.slot_number)
+        return earliest_slashed_epoch
+
+    def get_validators_with_earliest_exit_epoch(self, validators):
+        if len(validators) == 0:
+            return []
+
+        sorted_validators = sorted(validators, key = lambda validator: (int(validator.validator.validator.exit_epoch)))
+        earliest_exit_epoch = sorted_validators[0].validator.validator.exit_epoch
+        return filter_validators_by_exit_epoch(sorted_validators, earliest_exit_epoch)
+
+    def calc_validator_slashed_epoch_from_state(self, validator):
+        exit_epoch = int(validator.validator.validator.exit_epoch)
+        withdrawable_epoch = int(validator.validator.validator.withdrawable_epoch)
+
+        exited_period = withdrawable_epoch - exit_epoch
+        is_slashed_epoch_undetectable = exited_period > MIN_VALIDATOR_WITHDRAWABILITY_DELAY
+        if is_slashed_epoch_undetectable:
+            return None
+
+        return withdrawable_epoch - EPOCHS_PER_SLASHINGS_VECTOR
+
+    def find_earliest_slashed_epoch(self, validators, ref_slot):
+        pubkeys = get_validators_pubkeys(validators)
+        withdrawable_epoch = min(get_validators_withdrawable_epochs(validators))
+
+        start_slot = 0
```

Comment:
Let's think about optimization here, requests for validator states are heavy. 
We had the following thoughts here that might help:
- There is no point in going any deeper than the last finalized request
- We can use `activation_epoch` of the validator since it could not slash before
- We can iterate through slots with a multiple of 8192 where possible (at high values). To make it easier for the node to get states. Read more about it here: https://lighthouse-book.sigmaprime.io/advanced_database.html